driver-core/src/test/functional/com/mongodb/client/model/expressions/DocumentExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,235 @@
+        assertExpression(
+                BsonDocument.parse("{a: 1, r: null}"), // map.put("r", null)
+                a1.setField("r", Expressions.ofNull()),
+                // MQL:
+                "{'$setField': {'field': 'r', 'input': {'$literal': {'a': 1}}, 'value': null}}");
+
+        // Replacing a field based on its prior value:
+        assertExpression(
+                BsonDocument.parse("{a: 3}"), // map.put("a", map.get("a") * 3)
+                a1.setField("a", a1.getInteger("a").multiply(3)),
+                // MQL:
+                "{'$setField': {'field': 'a', 'input': {'$literal': {'a': 1}}, 'value': "
+                        + "{'$multiply': [{'$getField': {'input': {'$literal': {'a': 1}}, 'field': 'a'}}, 3]}}}");
+
+        // Placing a field based on a nested object:
+        assertExpression(
+                BsonDocument.parse("{'a': {'x': 1, 'y': 2}, r: 10}"),
+                ax1ay2.setField("r", ax1ay2.getDocument("a").getInteger("x").multiply(10)),
+                // MQL:
+                "{'$setField': {'field': 'r', 'input': {'$literal': {'a': {'x': 1, 'y': 2}}}, "
+                        + "'value': {'$multiply': ["
+                        + "  {'$getField': {'input': {'$getField': {'input': {'$literal': {'a': {'x': 1, 'y': 2}}}, "
+                        + "  'field': 'a'}}, 'field': 'x'}}, 10]}}}");
+
+        // Replacing a nested object requires two setFields, as expected:
+        assertExpression(
+                // "with" syntax: [ { a:{x:1,y:2} } ].map(d -> d.with("a", d.a.with("y", d.a.y.multiply(11))))
+                BsonDocument.parse("{'a': {'x': 1, 'y': 22}}"),
+                ax1ay2.setField("a", ax1ay2.getDocument("a")
+                        .setField("y", ax1ay2.getDocument("a").getInteger("y").multiply(11))),
+                "{'$setField': {'field': 'a', 'input': {'$literal': {'a': {'x': 1, 'y': 2}}}, "
+                        + "'value': {'$setField': {'field': 'y', 'input': {'$getField': "
+                        + "{'input': {'$literal': {'a': {'x': 1, 'y': 2}}}, 'field': 'a'}}, "
+                        + "'value': {'$multiply': [{'$getField': {'input': {'$getField': "
+                        + "{'input': {'$literal': {'a': {'x': 1, 'y': 2}}}, 'field': 'a'}}, "
+                        + "'field': 'y'}}, 11]}}}}}");
+    }
+
+    @Test
+    public void unsetFieldTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/unsetField/
+        assertExpression(
+                BsonDocument.parse("{}"), // map.remove("a")
+                a1.unsetField("a"),
+                // MQL:
+                "{'$unsetField': {'field': 'a', 'input': {'$literal': {'a': 1}}}}");
+    }
+
+    @Test
+    public void mergeTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/mergeObjects/
+        assertExpression(
+                BsonDocument.parse("{a: 1, b: 2}"),
+                ofDoc("{a: 1}").merge(ofDoc("{b: 2}")),
```

Comment:
Unsure whether this method should be on a record, since it would just take the right item. Makes more sense for when "objects" (documents) are being treated as maps (future PR).

Reply:
I don't understand this comment. Could you please elaborate?

Reply:
Will delete; needs to be on the Map type.

Reply:
Since the expressions API makes a distinction between a map (a variable set of field keys mapped to field values which are all of the same type) and a record (a arbitrary but fixed set of field keys mapped to field values of not necessary the same type), which are mixed within the same type "object"/"document" in MQL expressions, maybe we should consider not using the word "document" when referring to records, and just instead use the word "record"? I foresee that referring to records as documents may cause confusion (I was confused until we discussed this).

Reply:
I agree. But the renaming should be done at the end because of rebasing. https://github.com/mongodb/mongo-java-driver/pull/1032/files#r1032699544